lib/logstash/inputs/elasticsearch.rb

```diff
@@ -262,22 +268,39 @@ def run(output_queue)
   end
 
   private
-
+  JOB_NAME = "run query"
   def do_run(output_queue)
     # if configured to run a single slice, don't bother spinning up threads
-    return do_run_slice(output_queue) if @slices.nil? || @slices <= 1
+    return retryable(JOB_NAME) do
+      do_run_slice(output_queue)
+    end if @slices.nil? || @slices <= 1
 
     logger.warn("managed slices for query is very large (#{@slices}); consider reducing") if @slices > 8
 
     pipeline_id = execution_context&.pipeline_id || 'main'
     @slices.times.map do |slice_id|
       Thread.new do
         LogStash::Util::set_thread_name("[#{pipeline_id}]|input|elasticsearch|slice_#{slice_id}")
-        do_run_slice(output_queue, slice_id)
+        retryable(JOB_NAME) do
+          do_run_slice(output_queue, slice_id)
+        end
       end
     end.map(&:join)
   end
 
+  def retryable(job_name, &block)
+    begin
+      stud_try = ::LogStash::Helpers::LoggableTry.new(logger, job_name)
+      stud_try.try(@tries.times) {
```

Comment:
Thanks @kaisecheng for taking care this. This will be long lasting task and will give benefit alot!
One quick note at the glance: As I remember, `LogStash::Helpers::LoggableTry` tries on `StandardError` types. Do we need to specify timeout exceptions to make sure it retries only expected cases?
Eg.: `::LogStash::Outputs::ElasticSearch::HttpClient::Pool::HostUnreachableError`

Reply:
Yes, it catches `StandardError`. As the inputworker() fails to catch and restart plugin, any exception would crash Logstash which is the thing I want to avoid. Users running with many pipelines frustrate with the sudden shutdown. So I would say we need to have broader safeguard here.